pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceTransaction.java

```diff
@@ -0,0 +1,637 @@
+        Semaphore testThreadLimit = new Semaphore(arguments.numTestThreads);
+
+        long startTime = System.nanoTime();
+        long testEndTime = startTime + (long) (arguments.testTime * 1e9);
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            if (arguments.isEnableTransaction) {
+                printTxnAggregatedThroughput(startTime);
+            } else {
+                printAggregatedThroughput(startTime);
+            }
+            printAggregatedStats();
+        }));
+
+        // start perf test
+        AtomicBoolean executing = new AtomicBoolean(true);
+        new Thread(() -> {
+            RateLimiter rateLimiter = null;
+            if(arguments.openTxnRate != 0){
+                rateLimiter = RateLimiter.create( (double)arguments.openTxnRate / arguments.numTestThreads);
+            }
+            for(int i = 0; i < arguments.numTestThreads; i++) {
+                if (testThreadLimit.tryAcquire() && (rateLimiter == null || rateLimiter.tryAcquire())) {
+                    LongAdder messageSend = new LongAdder();
+                    LongAdder messageReceived = new LongAdder();
+                    executorService.submit(() -> {
+                        //The producer and consumer clients are built in advance, and then this thread is
+                        //responsible for the production and consumption tasks of the transaction through the loop.
+                        //A thread may perform tasks of multiple transactions in a traversing manner.
+                        List<Producer<byte[]>> producers = null;
+                        List<List<Consumer<byte[]>>> consumers = null;
+                        try {
+                            producers = buildProducers(client, arguments);
+                            consumers = buildConsumer(client, arguments);
+                        } catch (Exception e) {
+                            log.error("Failed to build Producer/Consumer with exception : " + e);
+                        }
+                        AtomicReference<Transaction> atomicReference = buildTransaction(client, arguments);
+                        //The while loop has no break, and finally ends the execution through the shutdownNow of
+                        //0the executorService
+                        while (true) {
+                            try {
+                                Transaction transaction = atomicReference.get();
+                                for (List<Consumer<byte[]>> subscriptions : consumers) {
+                                    if(messageReceived.sum() == arguments.numMessagesReceivedPerTransaction){
+                                        break;
+                                    }
+                                    for (Consumer<byte[]> consumer : subscriptions) {
+                                        if(messageReceived.sum() == arguments.numMessagesReceivedPerTransaction){
```

Comment:
every consumer receive `numMessagesReceivedPerTransaction` number messages and ack, this logic is not correct.

Reply:
Is the number of messages sent or received is not fixed in a transaction? The messageReceived here is a test thread, reset when the transaction is over, a test thread only executes one transaction at the same time.